include/crm/common/rules.h

```diff
@@ -48,6 +52,55 @@ enum expression_type {
 };
 //!@}
 
+//! Data used to evaluate a rule (any NULL items are ignored)
+typedef struct pcmk_rule_data {
+    // Used to evaluate date expressions
+    const crm_time_t *now; //!< Current time for rule evaluation purposes
+
+    // Used to evaluate resource type expressions
+    const char *rsc_standard;   //!< Resource standard that rule applies to
+    const char *rsc_provider;   //!< Resource provider that rule applies to
+    const char *rsc_agent;      //!< Resource agent that rule applies to
+
+    // Used to evaluate operation type expressions
+    const char *op_name;    //! Operation name that rule applies to
+    guint op_interval_ms;   //! Operation interval that rule applies to
+
+    // Remaining members are used to evaluate node attribute expressions
+
+    /*!
+     * Node attributes for rule evaluation purposes
+     *
+     * \note Though not const, this is used only with g_hash_table_lookup().
+     */
+    GHashTable *attrs;
+
+    // Remaining members are used only within location constraint rules
+
+    /*!
+     * Resource parameters that can be used as the reference value source
+     *
+     * \note Though not const, this is used only with g_hash_table_lookup().
+     */
+    GHashTable *rsc_params;
+
+    /*!
+     * Resource meta-attributes that can be used as the reference value source
+     *
+     * \note Though not const, this is used only with g_hash_table_lookup().
+     */
+    GHashTable *rsc_meta;
+
+    //! Resource ID to compare against a location constraint's resource pattern
+    const char *rsc_id;
+
+    //! Resource pattern submatches (as set by regexec()) for rsc_id
+    const regmatch_t *rsc_id_submatches;
+
+    //! Number of entries in rsc_id_submatches
+    int rsc_id_nmatches;
+} pcmk_rule_data_t;
```

Comment:
I went back and forth on this structure a few times. I considerd `pcmk_rule_input_t` and `pcmk_rule_evaluand_t` as names, but I didn't think any of the names was particularly intuitive, so I went with the shortest.

I like a single structure replacing five questionably differentiated structures. However some functions use only some of the members, which can be confusing too, so another possibility would be separate structs for resource expression evaluation (like `pe_rsc_eval_data_t`), operation expression evaluation (`pe_op_eval_data_t`), and attribute expression evaluation (`pe_re_match_data_t` + `pe_match_data_t` + part of `pe_rule_eval_data_t`). I didn't think it was worth the complexity.

Reply:
Groups of functions that share a particular interface (like callbacks) don't always use all their arguments, so I don't think it's a big deal if some functions ignore some members of `pcmk_rule_data_t`. Name sounds good to me too, all things considered, though `pcmk_rule_input_t` might be better.

Surprised it needs to be public but I haven't gotten far into looking at this.

Reply:
I plan on keeping a single public API for evaluating a rule (not in this PR)